Integrate `ProductSelectorFragment` into Order creation

samiuelson opened the pull request:
#### Order creation - products multi-selection
Closes: https://github.com/woocommerce/woocommerce-android/issues/8420

### Description
* Integrated `ProductSelectorFragment` into Order creation - product/variation multi-selection flow. The old solution allowed the selection of a single product. The new one provides a multi-selection of items.
* Deleted old product and variation picker code
* Removed part of Order creation UI test related to the old, deleted product selection code.

### Testing instructions
Go to the Orders screen and create a new Order. Tap + Add product screen and verify the multi-selection adds products/variations correctly to the order. Important scenarios to test:

* Can select multiple products and variations at the same time
* Products and variations already added to the order are marked as selected when returning to the product picker page
* Unselecting previously selected products/variations should remove them from the order
* In case the selected product quantity was updated on the order page it should be preserved when opening the product picker and returning back to the order
* Tapping on the selected product on the Orders page should redirect to the Item Details screen and allow removing the product
* The coupon edition screen (product pickers) should remain working without any changes as well.

### Images/gif
<!-- Include before and after images or gifs when appropriate. -->

https://user-images.githubusercontent.com/4527432/223449344-2b167912-ee09-4ece-adf8-7aaee2b2c5c6.mp4



- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

We'll do it when merging a final PR to trunk.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

samiuelson committed:
Integrate `ProductSelectorFragment` into Order creation

Side effects:
Removed old product picker code:
- OrderCreateEditProductSelectionFragment.kt
- OrderCreateEditProductSelectionViewModel.kt
- OrderCreateEditProductSelectionViewModelTest.kt
- OrderCreateEditVariationSelectionFragment.kt
- OrderCreateEditVariationSelectionViewModel.kt
- OrderCreateEditVariationSelectionViewModelTest.kt

Updated affected unit tests

samiuelson committed:
Adjust UI tests

Removed the `OrderSelectProductScreen` since the related implementation has been removed and replaced by `ProductSelectorFragment`

samiuelson committed:
Add unit tests

jostnes commented on WooCommerce/src/androidTest/kotlin/com/woocommerce/android/e2e/tests/ui/OrdersUITest.kt:
```diff
@@ -57,9 +57,6 @@ class OrdersUITest : TestBase() {
                 .createFABTap()
                 .assertNewOrderScreen()
                 .updateOrderStatus(status)
-                .addProductTap()
```
Rather than removing the steps from the test (and removing the test-related methods in [UnifiedOrderScreen.kt](https://github.com/woocommerce/woocommerce-android/pull/8475/files#diff-37741575cb09e8957e61e18be16da9d453587f6062c6d027f09980629e2c94c6) and [OrderSelectProductScreen.kt](https://github.com/woocommerce/woocommerce-android/pull/8475/files#diff-9fd8908dbc17d83a6dcfb179dc902e26a8ddee1f03d086ee9e2b3c137db5d489)), I think it would be better to temporarily ignore this test while the work is still in progress (It is still a valid and valuable test). You can use the `@Ignore` annotation for this so it would look something like this:
```
...
    @Test
    @Ignore ("Ignored temporarily, product selection step needs to be updated. Add link to GH issue here")
    fun e2eCreateOrderTest() {
...
```
The issue can be added to the project board to be looked at later (at least before merging feature branch). WDYT?

samiuelson replied:
Yep, we can do it like this for now. Reverted with 0092c04466ceb46e9685422077ee3d6e8c65567f and ignored with f842f3439f32e9fde0447addb5ea7cc2d710e5e6.

jostnes replied:
Thank you for considering the change and updating it!

jostnes commented:
I looked at the build, and overall, the feature worked very well! Great job with the feature!

I added one comment related to UI test and have a few comments around the feature:
1. A small thing I noticed while testing is that when I remove a product from the Orders screen and quickly click Add product again, the just deleted product is still displayed. It happened a few times that it got a bit annoying, would it be possible to disable the Add Product while delete is still in progress or are there other alternatives to make the experience better?

https://user-images.githubusercontent.com/17252150/223695086-abd0d9ef-16a8-41d8-b73a-81899fbd67fa.mp4

2. This is an odd one; on this branch, seems like more products are displayed (some of the products are draft), below are screenshots from the current beta build vs this build:
<img src="https://user-images.githubusercontent.com/17252150/223697183-1e659207-35ce-4c9e-90f7-48d0d9737f41.png" width="75%" height="75%" />

3. This is the same behavior on Coupon Management, on landscape mode. When the filter returned no results, the user would need to scroll to view the "clear filter" button, would it be possible to display it on screen without scrolling when in landscape mode (also a UI defect on the header, it's cut off on the top left)?
<img src="https://user-images.githubusercontent.com/17252150/223695811-2f16a550-af46-452b-a965-9790dd136e1a.png" width="50%" height="50%" />

samiuelson committed:
Disable "+ Add product" button after deleting item

It will be enabled again after the order is syced

samiuelson commented:
Thanks, @jostnes, for testing this PR! 🙌🏼 

Re 1. I've pushed the fix for that with https://github.com/woocommerce/woocommerce-android/pull/8475/commits/404e007501172ef68cefd68e520c4a4175781fe3.

Re 2. I've created an [issue](https://github.com/woocommerce/woocommerce-android/projects/49#card-88402813) and added it to the project board. We'll tackle it in the next PR.

Re 3. Also added to the board as a separate [issue](https://github.com/woocommerce/woocommerce-android/projects/49#card-88403019).

Regarding the UI tests, what is the comment you mentioned? Can't find it. FYI, I plan to explore and introduce Compose UI tests for product multi-selection as my HACK week project. I'll keep you up to date.

jostnes commented:
> Regarding the UI tests, what is the comment you mentioned? Can't find it. FYI, I plan to explore and introduce Compose UI tests for product multi-selection as my HACK week project. I'll keep you up to date.

first off, thanks for addressing the issues! as for the UI test comment, sorry, i wrote the comment but it was left at "pending" oops! i've just (really this time) submitted the comment, so you should be able to see it now

samiuelson committed:
Revert "Adjust UI tests"

This reverts commit 6d33390caea0eb8335ea6eaece63164492ca59a4.

samiuelson committed:
Ignore out-dated UI test

AnirudhBhat commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/creation/OrderCreateEditViewModel.kt:
```diff
@@ -228,27 +231,60 @@ class OrderCreateEditViewModel @Inject constructor(
             mapOf(KEY_FLOW to flow)
         )
-        it.adjustProductQuantity(item.itemId, -item.quantity.toInt())
+        viewState = viewState.copy(isEditable = false)
```
❓ What do you think about writing a unit test to verify the view state will not be editable while removing the product?

samiuelson replied:
Added with eda998ead15ae17ede61fecc2e0c48a052962e5f 👍

AnirudhBhat reviewed (approved):
Great work @samiuelson! I've tested the functionality and it works as expected. I've just left one minor comment that is non-blocking for the merge.

samiuelson committed:
Add unit test for view isEditable states
